### PR TITLE
Fix some bugs about search in code editor

### LINKED ISSUE #33315: "Replace" doesn't take effect on current text if selected from left to right
**Godot version:**
3.2 alpha3

**Steps to reproduce:**
1. Write some repeating script text (in new lines)
Like
```
var a = "whatever"
var b = "whatever"
var c = "whatever"
```
2. Select a repeating word
3. Ctrl + R to bring up replace
4. Write something and replace the text
5. It doesn't replace the first one, but jumps to the second

This only happens if you select the text from left to right (so the cursor is at the end of word). If you select in opposite way, there's one "blank" replace action on first click that does nothing

### LINKED ISSUE #61713: Ctrl-F find results in "-1 of n matches"
### Godot version

4.0.alpha9

### System information

Windows 10

### Issue description

Ctrl-F find can result in messages being displayed such as "-1 of 2 matches", when the correct behaviour should be "2 of 2 matches". This occurs when the search box is closed with text in it and then reopened without editing the contents of the search box.
image

### Steps to reproduce

1. Open the following script in the godot editor 
```
extends Node


# Called when the node enters the scene tree for the first time.
func _ready():
	pass # Replace with function body.


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta):
	pass
```
2. Ctrl-F and search for the `#` character
3. Hit Escape to close the search box
4. Ctrl-F again to reopen the search box.
5. Shift-Enter to go to the last match, "-1 of 3 matches" is displayed.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by Rindbee**
1. Fix #61713, Supersede #61715;
2. Fix the bug when there are consecutive matches, forward searching will skip the adjacent item;
3. Fix the bug that enable the selection-only option will affect the operations in search mode.
4. Fix #33315.

| Before | After |
| :------: | :------: |
|  |  |

When popup the search bar without a selection, just display the number of matches.

| Before | After |
| :------: | :------: |
|  |  |

**known problem**: In some extreme cases, searching forward and searching backward will give different results.

### CODE REVIEW

**Comment by KoBeWi** [comment]
Seems like it fixes #33315

**Comment by Rindbee** [comment]
Well, now when there is a selection, searching and replacing is performed from the beginning of the selection.

**Comment by Rindbee** [comment]
OK, now please have a review on it.

File: editor/code_editor.cpp [hunk 1]
```diff
@@ -391,10 +410,10 @@ void FindReplaceBar::_update_matches_label() {
 
 		if (results_count == 0) {
 			matches_label->set_text("No match");
-		} else if (results_count == 1) {
-			matches_label->set_text(vformat(TTR("%d match"), results_count));
+		} else if (results_count_to_current == -1) {
+			matches_label->set_text(vformat(results_count == 1 ? TTR("%d match") : TTR("%d matches"), results_count));
```
Comment (KoBeWi) [inline editor/code_editor.cpp:414]: I think this can be replaced with plural TTRN.

**Review by KoBeWi — APPROVED** [review]
Works correctly and the code looks ok, but needs a check from @Paulb23

File: editor/code_editor.cpp (hunk 1 above)
Comment (Rindbee) [reply editor/code_editor.cpp]: Thanks for pointing out.

File: editor/code_editor.h [hunk 2]
```diff
@@ -92,7 +92,7 @@ class FindReplaceBar : public HBoxContainer {
 	bool replace_all_mode = false;
 	bool preserve_cursor = false;
 
-	void _get_search_from(int &r_line, int &r_col);
+	void _get_search_from(int &r_line, int &r_col, bool is_searching_next = false);
```
Comment (akien-mga) [inline editor/code_editor.h:95]: Should be `p_is_searching_next`.

Comment (Rindbee) [reply editor/code_editor.h]: Thanks for pointing out.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**